cmd/client-admin.go

```diff
@@ -67,8 +71,74 @@ func NewAdminFactory() func(config *Config) (*madmin.AdminClient, *probe.Error)
 		var api *madmin.AdminClient
 		var found bool
 		if api, found = clientCache[confSum]; !found {
+			var credsChain []credentials.Provider
+
+			var transport http.RoundTripper
+
+			if config.Transport != nil {
+				transport = config.Transport
+			} else {
+				tr := &http.Transport{
+					Proxy:                 http.ProxyFromEnvironment,
+					DialContext:           newCustomDialContext(config),
+					MaxIdleConnsPerHost:   1024,
+					WriteBufferSize:       32 << 10, // 32KiB moving up from 4KiB default
+					ReadBufferSize:        32 << 10, // 32KiB moving up from 4KiB default
+					IdleConnTimeout:       90 * time.Second,
+					TLSHandshakeTimeout:   10 * time.Second,
+					ExpectContinueTimeout: 10 * time.Second,
+					DisableCompression:    true,
+				}
+				if useTLS {
+					// Keep TLS config.
+					tlsConfig := &tls.Config{
+						RootCAs: globalRootCAs,
+						// Can't use SSLv3 because of POODLE and BEAST
+						// Can't use TLSv1.0 because of POODLE and BEAST using CBC cipher
+						// Can't use TLSv1.1 because of RC4 cipher usage
+						MinVersion: tls.VersionTLS12,
+					}
+					if config.Insecure {
+						tlsConfig.InsecureSkipVerify = true
+					}
+					tr.TLSClientConfig = tlsConfig
+				}
+				transport = tr
+			}
+
+			transport = limiter.New(config.UploadLimit, config.DownloadLimit, transport)
+			// if an STS endpoint is set, we will add that to the chain
+			if stsEndpoint := env.Get("MC_STS_ENDPOINT", ""); stsEndpoint != "" {
+				// set AWS_WEB_IDENTITY_TOKEN_FILE is MC_WEB_IDENTITY_TOKEN_FILE is set
+				if val := env.Get("MC_WEB_IDENTITY_TOKEN_FILE", ""); val != "" {
+					os.Setenv("AWS_WEB_IDENTITY_TOKEN_FILE", val)
+				}
+
+				stsEndpointURL, err := url.Parse(stsEndpoint)
+				if err != nil {
+					return nil, probe.NewError(fmt.Errorf("Error parsing sts endpoint: %v", err))
+				}
+				credsSts := &credentials.IAM{
+					Client: &http.Client{
+						Transport: transport,
+					},
+					Endpoint: stsEndpointURL.String(),
+				}
+				credsChain = append(credsChain, credsSts)
+			}
+
+			// V4 Credentials
+			credsV4 := &credentials.Static{
+				Value: credentials.Value{
+					AccessKeyID:     config.AccessKey,
+					SecretAccessKey: config.SecretKey,
+					SessionToken:    config.SessionToken,
+					SignerType:      credentials.SignatureV4,
+				},
+			}
```

Comment:
We could make this a common function? only thing that needs to not be added is `signature.v2` which is not supported with Admin APIs.

Reply:
right, thats why I left v2 outside.

Should also reuse `transport` instead of overwrite it? there is a difference in couple of different settings:

```
MaxIdleConnsPerHost:   1024,
WriteBufferSize:       32 << 10, // 32KiB moving up from 4KiB default
ReadBufferSize:        32 << 10, // 32KiB moving up from 4KiB default
```

Reply:
> right, thats why I left v2 outside.
> 
> Should also reuse `transport` instead of overwrite it? there is a difference in couple of different settings:
> 
> ```
> MaxIdleConnsPerHost:   1024,
> WriteBufferSize:       32 << 10, // 32KiB moving up from 4KiB default
> ReadBufferSize:        32 << 10, // 32KiB moving up from 4KiB default
> ```

`mc` just runs and dies so there is nothing to re-use really. The transport is re-used always until `mc` is running.

and it is not created multiple times as such there is a `clientCache` of the `mc` config which is loaded per target in memory.

Reply:
Yes these settings are okay we should combine both transports just that add v2 optionally when its used by an S3 client.